addons/blocks2image/userscript.js

```diff
@@ -0,0 +1,233 @@
+
+
+    let svg = document.createElement('svg')
+    svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg')
+    svg.setAttribute('xmlns:html', 'http://www.w3.org/1999/xhtml')
+    svg.setAttribute('xmlns:xlink', 'http://www.w3.org/1999/xlink')
+    svg.setAttribute('version', '1.1')
+
+    let style = document.createElement('style')
+    style.textContent =
+        `
+.blocklyText {
+    fill: #fff;
+    font-family: "Helvetica Neue", Helvetica, sans-serif;
+    font-size: 12pt;
+    font-weight: 500;
+}
+.blocklyNonEditableText>text, .blocklyEditableText>text {
+    fill: #575E75;
+}
+.blocklyDropdownText {
+    fill: #fff !important;
+}
+`
+    if (request.command === 'export1' || request.command === 'export3') {
+        svg = selectedBlocks(svg, style, isExportPNG)
+    } else {
+        svg = allBlocks(svg, style, isExportPNG)
+    }
+    // resolve nbsp whitespace
+    let texts = Array.from(svg.getElementsByTagName('text'))
+    texts.forEach(text => {
+        text.innerHTML = text.innerHTML.replace(/&nbsp;/g, ' ')
+    })
+    // resolve image path
+    let images = Array.from(svg.getElementsByTagName('image'))
+    let scratchURL = window.location.origin
+
+    images.forEach(item => {
+        let builtinSvgData = blocksMedia.get(item.getAttribute('xlink:href').substring(item.getAttribute('xlink:href').lastIndexOf('/') + 1))
+        if (builtinSvgData) {   // replace svg file path (offical) to inline svg
+            item.setAttribute('xlink:href', builtinSvgData)
+        } else if (item.getAttribute('xlink:href').indexOf('/static/') === 0) {    // replace link path for third party website
+            item.setAttribute('xlink:href', scratchURL + item.getAttribute('xlink:href').slice(0))
+        } else if (item.getAttribute('xlink:href').indexOf('./static/') === 0) {
+            item.setAttribute('xlink:href', scratchURL + item.getAttribute('xlink:href').slice(1))
+        } else if (item.getAttribute('xlink:href').indexOf('static/') === 0) {
+            item.setAttribute('xlink:href', scratchURL + '/' + item.getAttribute('xlink:href'))
+        }
+    })
+    let tmp = document.createElement('div')
+    tmp.appendChild(svg);
+    if (request.command === 'export1' || request.command === 'export2') {
+        exportData(tmp.innerHTML)
+    } else {
+        exportPNG(svg);
+    }
+}
+
+function selectedBlocks(svg, style, isExportPNG) {
+    let svgchild = document.querySelector('svg.blocklySvg g.blocklySelected')
+    if (!svgchild) {
+        alert('Click on the blocks you want to export!')
```

Comment:
l10n the alert. You don't need to do that for the error though because that won't be shown to users.

Reply:
What does this mean? How do I simply alert the user to a problem with the operation? I see that there seems to be addon.notifications, is this the one to use?

Reply:
Add more l10n so the alert can be translated.

The function will have to be inside the main `export default async function` in order to have access to the `msg()` function, but that's not a problem. It won't break anything.

Reply:
I get it, thx